oxidation/libparsec/crates/serialization_format/src/protocol.rs

```diff
@@ -0,0 +1,739 @@
+                    // to be missing to still be compatible with APIv1.0)
+                    let added_in_minor_revision = match &field.introduced_in {
+                        Some(introduced_in) => {
+                            has_introduced_in_field.replace(true);
+                            match parse_api_version(introduced_in) {
+                                Ok((introduced_in_major, _)) => {
+                                    if introduced_in_major > major_version {
+                                        // Field has been added in a subsequent major version,
+                                        // hence our current major version doesn't know about it !
+                                        return None;
+                                    }
+                                    introduced_in_major == major_version
+                                }
+                                Err(err) => panic!("{:?}: {:?}", cmd.req.cmd, err),
+                            }
+                        }
+                        None => false,
+                    };
+                    Some(GenCmdField {
+                        name: field.name.clone(),
+                        ty: field.ty.clone(),
+                        added_in_minor_revision,
+                        allowed_extra_types: allowed_extra_types.clone(),
+                    })
+                };
+
+                let gen_nested_types = {
+                    if let Some(ref nested_types) = cmd.nested_types {
+                        nested_types.iter().map(|nested_type| {
+                            match (&nested_type.fields, &nested_type.variants, &nested_type.discriminant_field) {
+                                (None, Some(variants), discriminant_field) => {
+                                    GenCmdNestedType::Enum {
+                                        name: nested_type.name.to_owned(),
+                                        discriminant_field: discriminant_field.to_owned(),
+                                        variants: variants.iter().map(
+                                            |variant| {
+                                                GenCmdNestedTypeVariant {
+                                                    name: variant.name.to_owned(),
+                                                    discriminant_value: variant.discriminant_value.to_owned(),
+                                                    fields: variant.fields.as_ref().map(|fields| {
+                                                        fields.iter()
+                                                        .filter_map(&convert_field)
+                                                        .collect()
+                                                    }),
+                                                }
+                                            }
+                                        ).collect()
+                                    }
```

Comment:
This code use the `convert_field` mut closure so it's not easy to move it

I'd rather keep the parsing work in a single place as much as possible:
- it's easier to follow the flow
- there no need for factorization since this code is only needed in a single place
- it simplifies providing context to `panic!/assert!` (since the parent structs are available). Otherwise we would have either less context in errors (e.g. not the name of the command that contains an error in it field) or more complex code (passing context info, or returning a Result type and having to deal with it in the caller place).